File: sdk/python/tests/test_simple.py
"""Simple SDK test: create a trace, add a span, complete it, verify."""

import os
import sys

# Allow running without installing: add the SDK to sys.path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from llmtrace import LLMTrace, CustomKind

BASE = os.environ.get("LLMTRACE_URL", "http://localhost:3000")


def test_create_and_complete_span():
    """Round-trip: create trace → create span → complete → verify."""
    client = LLMTrace(BASE)

    # Clean slate
    client.clear_all()

    # Create a trace
    with client.trace("simple-test") as t:
        trace_id = t.trace_id
        assert trace_id, "trace_id should be set"

        # Create a span inside the trace
        with t.span("step-1", kind=CustomKind(kind="task", attributes={"key": "val"})) as s:
            s.set_output({"answer": 42})

    # Verify trace exists
    traces = client.get_traces()
    assert traces.count >= 1
    found = any(tr.id == trace_id for tr in traces.traces)
    assert found, f"Trace {trace_id} not in trace list"

    # Verify span
    span_list = client.get_trace(trace_id)
    assert span_list.count == 1
    span = span_list.spans[0]
    assert span.name == "step-1"
    assert span.status == "completed"
    assert span.output == {"answer": 42}
    assert isinstance(span.kind, CustomKind)
    assert span.kind.kind == "task"

    # Verify stats
    stats = client.get_stats()
    assert stats.span_count >= 1

    # Cleanup
    client.delete_trace(trace_id)
    client.close()
    print("PASS: test_create_and_complete_span")


def test_fail_span():
    """A span that raises should be marked as failed."""
    client = LLMTrace(BASE)

    with client.trace("fail-test") as t:
        trace_id = t.trace_id
        try:
            with t.span("will-fail", kind=CustomKind(kind="task", attributes={})) as s:
                raise ValueError("something broke")
        except ValueError:
            pass  # expected

    span_list = client.get_trace(trace_id)
    assert span_list.count == 1
    span = span_list.spans[0]
    assert span.status == "failed"
    assert span.error == "something broke"

    client.delete_trace(trace_id)
    client.close()
    print("PASS: test_fail_span")


if __name__ == "__main__":
    test_create_and_complete_span()
    test_fail_span()
    print("\nAll simple tests passed!")

File: sdk/typescript/tests/test_simple.ts
/**
 * Simple SDK test: create a trace, add a span, complete it, verify.
 *
 * Run: npx tsx tests/test_simple.ts
 * (from sdk/typescript/, no build step needed)
 */

import { Traceway, type SpanKind } from '../src/index.js';
import { statusKind, statusError } from '../src/types.js';

const BASE = process.env.TRACEWAY_URL ?? 'http://localhost:3000';
const client = new Traceway({ url: BASE });

function assert(cond: boolean, msg: string) {
  if (!cond) throw new Error(`ASSERT FAILED: ${msg}`);
}

async function testCreateAndCompleteSpan() {
  // Clean slate
  await client.clearAll();

  // Create a trace with a span
  const traceId = await client.trace('simple-test', async (t) => {
    await t.span('step-1', async (s) => {
      s.setOutput({ answer: 42 });
    }, {
      kind: { type: 'custom', kind: 'task', attributes: { key: 'val' } },
    });
    return t.traceId;
  });

  // Verify trace exists
  const traces = await client.getTraces();
  assert(traces.count >= 1, `Expected >=1 traces, got ${traces.count}`);
  const found = traces.traces.some(tr => tr.id === traceId);
  assert(found, `Trace ${traceId} not in trace list`);

  // Verify span
  const spanList = await client.getTrace(traceId);
  assert(spanList.count === 1, `Expected 1 span, got ${spanList.count}`);
  const span = spanList.spans[0];
  assert(span.name === 'step-1', `Expected name 'step-1', got '${span.name}'`);
  assert(statusKind(span.status) === 'completed', `Expected completed, got ${JSON.stringify(span.status)}`);
  assert(JSON.stringify(span.output) === JSON.stringify({ answer: 42 }), 'Output mismatch');

  // Verify kind
  assert(span.kind?.type === 'custom', `Expected custom kind, got ${span.kind?.type}`);

  // Verify stats
  const stats = await client.getStats();
  assert(stats.span_count >= 1, `Expected >=1 spans, got ${stats.span_count}`);

  // Cleanup
  await client.deleteTrace(traceId);
  console.log('PASS: testCreateAndCompleteSpan');
}

async function testFailSpan() {
  const traceId = await client.trace('fail-test', async (t) => {
    try {
      await t.span('will-fail', async () => {
        throw new Error('something broke');
      }, {
        kind: { type: 'custom', kind: 'task', attributes: {} },
      });
    } catch {
      // expected
    }
    return t.traceId;
  });

  const spanList = await client.getTrace(traceId);
  assert(spanList.count === 1, `Expected 1 span, got ${spanList.count}`);
  const span = spanList.spans[0];
  assert(statusKind(span.status) === 'failed', `Expected failed, got ${JSON.stringify(span.status)}`);

  const err = statusError(span.status);
  assert(err !== null && err.includes('something broke'), `Expected error containing 'something broke', got '${err}'`);

  await client.deleteTrace(traceId);
  console.log('PASS: testFailSpan');
}

async function main() {
  await testCreateAndCompleteSpan();
  await testFailSpan();
  console.log('\nAll simple tests passed!');
}

main().catch(e => {
  console.error(e);
  process.exit(1);
});
